crates/uv-auth/src/middleware.rs

```diff
@@ -48,6 +49,54 @@ impl NetrcMode {
     }
 }
 
+#[derive(Copy, Clone, Debug, Default, Hash, Eq, PartialEq)]
+pub enum AuthPolicy {
+    /// Attempt to authenticate if needed.
+    #[default]
+    Auto,
+    /// Always attempt to authenticate.
+    Always,
+    /// Never attempt to authenticate.
+    Never,
+}
+
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub struct UrlAuthPolicies(FxHashMap<Url, AuthPolicy>);
+
+impl UrlAuthPolicies {
+    pub fn new() -> Self {
+        Self(FxHashMap::default())
+    }
+
+    /// Create a new `UrlAuthPolicies` from a list of URL and `AuthPolicy`
+    /// tuples.
+    pub fn from_tuples(tuples: &[(Url, AuthPolicy)]) -> Self {
+        let mut auth_policies = UrlAuthPolicies::new();
+        for (url, auth_policy) in tuples {
+            auth_policies.add_policy(url.clone(), *auth_policy);
```

Comment:
As a rule of thumb, if you're taking unowned data but cloning all of it, I would often suggest taking an owned value. Otherwise, you're kind of hiding the allocations from the caller.